feat: Search sidebar

TomWoodward opened the pull request.

Commit:
Started toggle functinality for search bar results and styles for when it's visible

Commit:
Styling for search results on bar when there are actual results

Commit:
styles for chapters in results

Commit:
Merge remote-tracking branch 'origin/master' into search-sidebar

Commit:
Merge remote-tracking branch 'origin/master' into search-sidebar

Commit:
selector and initial styles based on real results

Commit:
changed how hits from search are accessed

TomWoodward committed:
fix sidebar props

Commit:
first attempt to fix left-margin on entire search results bar

TomWoodward committed:
Merge branch 'master' into search-sidebar

TomWoodward commented on src/app/content/components/SearchResultsSidebar.tsx:
```diff
@@ -0,0 +1,278 @@
+`;
+
+// tslint:disable-next-line:variable-name
+const SearchResultsBar = styled.div`
```
most of this is the same as the ToC wrapper, we should try to combine as much as possible

TomWoodward commented on src/app/content/components/SearchResultsSidebar.tsx:
```diff
@@ -0,0 +1,278 @@
+  }
+
+  &[open] {
+    display: block;
```
the sidebar has a transition that looks nicer than this, also use props instead of html attributes for this kind of thing

TomWoodward commented on src/app/content/components/SearchResultsSidebar.tsx:
```diff
@@ -0,0 +1,278 @@
+  ${theme.breakpoints.mobile(css`
+    width: calc(50vw + ${searchResultsBarMobileWidth}vw);
+    min-width: calc(50vw + ${searchResultsBarMobileWidth}vw);
+    top: ${bookBannerMobileMiniHeight + toolbarMobileHeight
```
we're gonna have to give this access to whether the search input is open on mobile and compensate so the top position is right whether its open or closed

probably do the same in the toc sidebar

thats sort of a pain...

TomWoodward commented on src/app/content/components/SearchResultsSidebar.tsx:
```diff
@@ -0,0 +1,278 @@
+    min-width: calc(50vw + ${searchResultsBarMobileWidth}vw);
+    top: ${bookBannerMobileMiniHeight + toolbarMobileHeight
+          + toolbarSearchInputMobileHeight + (mobileSearchContainerMargin * 2)}rem;
+    height: calc(100vh - ${navMobileHeight + bookBannerMobileMiniHeight}rem);
```
this height looks wrong just because its less complicated than the top calculation

TomWoodward commented on src/app/content/components/SearchResultsSidebar.tsx:
```diff
@@ -0,0 +1,278 @@
+    top: ${bookBannerMobileMiniHeight + toolbarMobileHeight
+          + toolbarSearchInputMobileHeight + (mobileSearchContainerMargin * 2)}rem;
+    height: calc(100vh - ${navMobileHeight + bookBannerMobileMiniHeight}rem);
+    margin-left: calc(-50vw - ${sidebarMobileWidth}rem - ${searchResultsBarVariables.mainPaddingMobile}rem);
```
i don't think you need this `searchResultsBarVariables.mainPaddingMobile` part here

TomWoodward commented on src/app/content/components/SearchResultsSidebar.tsx:
```diff
@@ -0,0 +1,278 @@
+  transition: transform 300ms ease-in-out,box-shadow 300ms ease-in-out,background-color 300ms ease-in-out;
+  background-color: #fafafa;
+  z-index: 4;
+  margin-left: calc(-50vw - ${sidebarDesktopWidth}rem - ${searchResultsBarVariables.mainPaddingDesktop}rem);
```
i don't think you need this `searchResultsBarVariables.mainPaddingMobile` part here

TomWoodward commented on src/app/content/components/constants.ts:
```diff
@@ -9,10 +9,17 @@ export const sidebarDesktopWidth = 33.5;
 
+export const mobileSearchContainerMargin = 1;
+export const toolbarSearchInputDesktopHeight = 3.2;
+export const toolbarSearchInputMobileHeight = 3.2;
```
if we also export variables for `toolbarSearchDesktopHeight` `toolbarSearchMobileOpenHeight` `toolbarSearchMobileClosedHeight` that might make logic elsewhere simpler

TomWoodward commented on src/app/content/components/SearchResultsSidebar.tsx:
```diff
@@ -0,0 +1,278 @@
+
+  ${theme.breakpoints.mobile(css`
+    width: calc(50vw + ${searchResultsBarMobileWidth}vw);
+    min-width: calc(50vw + ${searchResultsBarMobileWidth}vw);
```
your search bar width should not be in `vw`

TomWoodward reviewed (commented):
looking good so far, the only big thing missing is that there is supposed to be a loading state while the query is processing using the bouncy openstax logo from osweb

Commit:
loading state initiated

Commit:
hidden loading state

TomWoodward committed:
remove extra padding on search bar

TomWoodward committed:
make close button work

Commit:
open attribute no longer required

Commit:
Merge branch 'search-sidebar' of https://github.com/openstax/rex-web into search-sidebar

TomWoodward committed:
Merge branch 'search-sidebar' of github.com:openstax/books-web into search-sidebar

TomWoodward committed:
reset results when submitting new searches

Commit:
loading state based on results

Commit:
merge conflicts

Commit:
making close icon a button

Commit:
search results bar width on mobile

Commit:
unused import

Commit:
margin top on mobile search results  bar

TomWoodward committed:
Merge remote-tracking branch 'origin/master' into search-sidebar

TomWoodward committed:
Merge branch 'search-sidebar' of github.com:openstax/books-web into search-sidebar

TomWoodward committed:
hook up real results

Commit:
vertical alignment for long search query

Commit:
merge conflicts

Commit:
style fixes on search bar for desktop

Commit:
height fix with scrolling, fix for when
 no results, mobile width fix, moved
link to description in results

TomWoodward committed:
Merge remote-tracking branch 'origin/master' into search-sidebar

Commit:
Merge remote-tracking branch 'origin/master' into search-sidebar

Commit:
separated things into different files,
scrolling handler and initiated toggle functionality.

Commit:
variable modifications

TomWoodward committed:
working on sidebar display

Commit:
toggling search results in all scenarios
controlling visibility on Layout

TomWoodward commented on src/app/components/Loader.tsx:
```diff
@@ -0,0 +1,124 @@
+
+const LoadingIcon = styled.svg`
+  enable-background:new 0 0 57.6 39.1;
+                animation: ${loadSvg} 0.3s forwards 0.3s;
```
whitespace

TomWoodward commented on src/app/content/components/SearchResultsSidebar/index.tsx:
```diff
@@ -0,0 +1,94 @@
+import * as Styled from './styled';
+
+// tslint:disable-next-line:variable-name
+export const SearchResultContainers = (props: {containers: SearchResultContainer[],
```
these aren't used in this file, we should probably move them to separate files

TomWoodward commented on src/app/content/components/SearchResultsSidebar/index.tsx:
```diff
@@ -0,0 +1,94 @@
+import { SearchResultHit } from '@openstax/open-search-client/dist/models/SearchResultHit';
```
lets move these search sidebar files into `/search/components/`

TomWoodward commented on src/app/content/components/SearchResultsSidebar/SearchResultsBarWrapper.tsx:
```diff
@@ -0,0 +1,83 @@
+  public render() {
+    const {query, totalHits, results, onClose, book, open, toggleSearchSidebar} = this.props;
+
+    return <Styled.SearchResultsBar ref={this.searchSidebar} open={open}>
```
why does this have to have the `open` prop? won't it always be true?

TomWoodward commented on src/app/content/components/SearchResultsSidebar/SearchResultsBarWrapper.tsx:
```diff
@@ -0,0 +1,83 @@
+
+    window.addEventListener('scroll', animation, {passive: true});
+    window.addEventListener('resize', animation, {passive: true});
+    scrollHandler();
```
try to move this logic into a util function to be shared with the `scrollHandler` in the ToC

TomWoodward commented on src/app/content/components/SearchResultsSidebar/styled.tsx:
```diff
@@ -0,0 +1,210 @@
+    backgroundColor: '#f1f1f1',
+    iconRightPadding: 0.7,
+    mainPaddingDesktop: 3,
+    mainPaddingMobile: 2,
```
is this supposed to be the same as the page padding in the app theme?

KarinaMendez2 replied:
probably, but I followed the invision design, says 30px, already removed the mainPaddingMobile that was unused.

TomWoodward commented on src/app/content/components/SearchResultsSidebar/styled.tsx:
```diff
@@ -0,0 +1,210 @@
+import { toolbarIconStyles } from '../Toolbar';
+
+const searchResultsBarVariables = {
+    backgroundColor: '#f1f1f1',
```
can get this color from the app theme colors

TomWoodward commented on src/app/content/components/Toolbar.tsx:
```diff
@@ -3,22 +3,29 @@ import React from 'react';
 import { TimesCircle } from 'styled-icons/fa-solid/TimesCircle';
 import { maxNavWidth } from '../../components/NavBar';
-import { contentFont, textRegularLineHeight, textRegularSize, textRegularStyle } from '../../components/Typography';
+import { contentFont, textRegularLineHeight,
```
weird formatting here

TomWoodward commented on src/app/content/components/Toolbar.tsx:
```diff
@@ -3,22 +3,29 @@ import React from 'react';
 import { FormattedMessage } from 'react-intl';
 import { connect } from 'react-redux';
 import styled, { css } from 'styled-components/macro';
+import { ChevronLeft } from 'styled-icons/boxicons-regular/ChevronLeft';
```
use the font-awesome icons, not `boxicons`

TomWoodward commented on src/app/content/components/Wrapper.tsx:
```diff
@@ -1,9 +1,24 @@
   overflow: visible; /* so sidebar position: sticky works */
+  flex: 1;
+  ${(props) => {
+    if (props.open) {
```
this could use a more descriptive name like `searchOpen` or something, and can be reformatted like

```
styled(asdf)`
  ${(props) => props.open && css`
    display: none;
  `}
`;

TomWoodward commented on src/app/content/components/constants.ts:
```diff
@@ -9,10 +9,24 @@ export const sidebarDesktopWidth = 33.5;
 export const sidebarTransitionTime = 300;
 
+export const searchResultsBarDesktopWidth = 43.5;
+export const searchResultsBarMobileWidth = 100;
```
unused

TomWoodward commented on src/app/content/components/constants.ts:
```diff
@@ -9,10 +9,24 @@ export const sidebarDesktopWidth = 33.5;
+export const toolbarSearchInputMobileWidth = 18.5;
+export const toolbarToggleSearchBarTextMobileWidth = 11.2;
+export const toolbarMobileSearchWrapperHeight = toolbarSearchInputMobileHeight
++ (mobileSearchContainerMargin * 2) + 0.1;
```
where does this `0.1` come from?

KarinaMendez2 replied:
From the <hr/> border height that separates the toolbar from the mobile search input

TomWoodward commented on src/app/content/components/constants.ts:
```diff
@@ -9,10 +9,24 @@ export const sidebarDesktopWidth = 33.5;
+export const toolbarToggleSearchBarTextMobileWidth = 11.2;
+export const toolbarMobileSearchWrapperHeight = toolbarSearchInputMobileHeight
++ (mobileSearchContainerMargin * 2) + 0.1;
+export const toolbarMobileElementsHeight = bookBannerMobileMiniHeight
```
```suggestion
export const searchSidebarTopOffset = bookBannerMobileMiniHeight
```

TomWoodward commented on src/app/content/search/reducer.ts:
```diff
@@ -7,17 +7,18 @@ import { State } from './types';
 
 export const initialState = {
   loading: false,
+  open: false,
```
might consider renaming this `mobileOpen` since its only respected on mobile

TomWoodward commented on src/app/content/search/reducer.ts:
```diff
@@ -7,17 +7,18 @@ import { State } from './types';
     }
     case getType(actions.receiveSearchResults): {
-      return {...state, loading: false, results: action.payload};
+      return {...state, loading: false, results: action.payload, open: true};
```
you shouldn't need to add `open: true` here right, because its already in `requestSearch`?

KarinaMendez2 replied:
right.

TomWoodward commented on src/app/content/search/reducer.ts:
```diff
@@ -27,6 +28,9 @@ const reducer: Reducer<State, AnyAction> = (state = initialState, action) => {
         : state;
     }
+    case getType(actions.toggleSearchSidebar): {
+      return {...state, open: action.payload};
```
if this is a toggle then it should be `return {...state, open: !state.open};`, if you need to be able to pass in the open/closed boolean to make something work then probably should create two actions `openResults` and `closeResults`

TomWoodward commented on src/app/content/components/SearchResultsSidebar/styled.tsx:
```diff
@@ -0,0 +1,210 @@
+export const HeaderQuery = styled.div`
+  max-width: calc(${searchResultsBarDesktopWidth - (textRegularLineHeight * 2)
+              - searchResultsBarVariables.mainPaddingDesktop
+              - searchResultsBarVariables.mainRightPaddingDesktop}rem)
```
this is complicated, why is this complicated?

KarinaMendez2 replied:
because i'm calculating the space for that div to start overlapping the text if it's a long search

TomWoodward commented on src/app/content/components/SearchResultsSidebar/index.tsx:
```diff
@@ -0,0 +1,94 @@
+  public render() {
+    const {query, open} = this.props;
+
+    return query && open && <SearchResultsBarWrapper {...this.props}/>;
```
we don't have spec for open/closing the sidebar on desktop, so the `open` should only be respected on mobile. meaning if you're on mobile and click a search result (closing the sidebar) then resize your screen up you should get the sidebar again.

i think this means we have to do the hiding of the sidebar when it is closed (but not cleared) in css rather than this react condition.

TomWoodward commented on src/app/content/components/Toolbar.tsx:
```diff
@@ -323,9 +389,15 @@ class Toolbar extends React.Component<{
+      dispatch(clearSearch());
+    },
     search: (query: string) => dispatch(requestSearch(query)),
+    toggleSearchSidebar: (open: boolean) => dispatch(toggleSearchSidebar(open)),
```
not sure if you only put this here for testing, but we don't want this here. when you click the magnifying glass we want to `clearSearch`, the only way to close the sidebar without clearing it is to click on one of the result links

KarinaMendez2 replied:
We need the "openSearchResults" action for the "< Search Results" button that will open the results bar.

TomWoodward replied:
right ok, but its not being used for that now

Commit:
lint errors, re formatting functions and
changinf toggleSearchSidebar into separate open and close actions

Commit:
lint errors, re organizing some styles,
moved searchsidebar files to /search/components and attempt to move scrollhandler function to domUtils

Commit:
expand active chapter on reload and
set current page

Commit:
scrollHandler callback function fix

Commit:
unused import

TomWoodward committed:
Merge remote-tracking branch 'origin/master' into search-sidebar

TomWoodward committed:
don't navigate when loading scroll on page load

Commit:
Merge remote-tracking branch 'origin/master' into search-sidebar

Commit:
updated specs, started new ones for
search sidebar and scroll to selected
page working

Commit:
screenshot specs for search sidebar

TomWoodward commented on src/app/content/search/components/SearchResultsSidebar/styled.tsx:
```diff
@@ -0,0 +1,208 @@
+  width: 100%;
+  height: 100%;
+  background-color: ${searchResultsBarVariables.backgroundColor};
+  z-index: 5;
```
is this necessary?

TomWoodward commented on src/app/content/search/components/SearchResultsSidebar/styled.tsx:
```diff
@@ -0,0 +1,208 @@
+  max-width: 26.5rem;
+  text-align: justify;
+  /*textRegularLineHeight is the close icon height*/
+  margin-top: calc(7rem - ${textRegularLineHeight}rem);
```
why do you have to pull this up by the height of the icon? maybe the icon should be absolutely positioned? or this should be a row? i forget what it looks like '

TomWoodward commented on src/app/content/search/components/SearchResultsSidebar/styled.tsx:
```diff
@@ -0,0 +1,208 @@
+export const HeaderQuery = styled.div`
+  max-width: calc(${searchResultsBarDesktopWidth - (textRegularLineHeight * 2)
+              - searchResultsBarVariables.mainPaddingDesktop
+              - searchResultsBarVariables.mainRightPaddingDesktop}rem)
```
not sure what is up with this, seems like it should be simpler

TomWoodward commented on src/app/content/search/hooks.spec.ts:
```diff
@@ -127,7 +127,7 @@ describe('hooks', () => {
       });
 
-      expect(dispatch).toHaveBeenCalledWith(requestSearch('qwer'));
+      expect(dispatch).not.toHaveBeenCalledWith(requestSearch('qwer'));
```
i don't think you were supposed to change this. `'searches for saved query on POP if it is differet from current query'` sounds right to me

KarinaMendez2 replied:
this test was failing, saw the response and changed it based on that, since we modified the thing that did a search on reload

TomWoodward commented on src/app/content/search/selectors.ts:
```diff
@@ -10,3 +11,19 @@ export const query = createSelector(
   (state) => state.query
 );
+
+export const totalHits = createSelector(
```
seems easier to get this off of results, since if there are results you will definitely have `results.hits.total`, by making another selector that might return null you just have one more null check you have to do

TomWoodward commented on src/app/theme.ts:
```diff
@@ -22,6 +22,10 @@ const padding = {
 };
 
+const searchSidebar = {
+  background: '#f1f1f1',
```
sigh... since this color is literally only used in the search results sidebar maybe we should just define it at the top of that file

TomWoodward commented on src/config.test.js:
```diff
@@ -5,6 +5,8 @@ module.exports = {
 
   BOOKS: {'testbook1-uuid':{'defaultVersion':'1.0'}},
 
+  SEARCH_URL: process.env.SEARCH_URL || 'https://search-jul11.sandbox.openstax.org',
```
you don't want this

TomWoodward commented on src/setupProxy.js:
```diff
@@ -68,6 +68,7 @@ const findFileIn = (baseDir, reqInfo) => {
   const queryFilePath = path.join(filePath,
     reqInfo.search ? encodeURIComponent(reqInfo.search) : ''
   );
+  console.log(queryFilePath);
```
debug code

TomWoodward commented on src/test/fixtures/open-search/api/v0/search/%3Fq%3Dlaksdkd%26books%3Dtestbook1-uuid%25401.0%26index_strategy%3Di1%26search_strategy%3Ds1:
```diff
@@ -0,0 +1 @@
+{"overall_took":114,"took":4,"timed_out":false,"_shards":{"total":1,"successful":1,"skipped":0,"failed":0},"hits":{"total":0,"hits":[]}}
```
can you format this json just to make it a little nicer to read?

KarinaMendez2 commented on src/app/content/components/Toolbar.tsx:
```diff
@@ -233,46 +244,99 @@ const TopBar = styled.div`
+
+// tslint:disable-next-line:variable-name
+const InnerText = styled.div`
+  width: 11.2rem;
```
this is the width of the text "search results", never changes

TomWoodward replied:
but it doesn't have to be explicitly set right? it'll figure it out based on the font? that way if we change the font styles we don't have to figure out what the new width is?

Commit:
fixes for functions shared and re organizing code

TomWoodward committed:
fix hiding results after clicking and search results text width

KarinaMendez2 commented on src/app/content/search/components/SearchResultsSidebar/index.tsx:
```diff
@@ -0,0 +1,47 @@
+  closeSearchResults: () => void;
+}
+
+export class SearchResultsSidebar extends Component<SearchResultsSidebarProps> {
```
We did this so that the ref in the SearchResultsBarWrapper would asume the element would never be null

TomWoodward committed:
fix imports in sidebarspec

TomWoodward committed:
fixing TOC tests

TomWoodward committed:
unit tests pass

TomWoodward committed:
coverage for toolbar

Commit:
replacing colors and numbers with variables, renaming functions and
elements, screenshot and browser specs.

Commit:
merge conflicts

TomWoodward committed:
Adding coverage for search components

TomWoodward committed:
Merge branch 'search-sidebar' of github.com:openstax/books-web into search-sidebar

TomWoodward committed:
:camera:

TomWoodward committed:
test coverage for search results sidebar

TomWoodward committed:
cover more complicated search results

Commit:
changind search sidebar size,
line height in elementes, box shadows and elemente alignments

Commit:
modified clearSearch on tocOpen functionality, more line-heights, paddings and box-shadow fixes

Commit:
safari fixes (line-height and list display)

Commit:
added value to submit button on search, updated specs

Commit:
probable fix for search input placeholder in firefox

Commit:
merge conflicts, snapshots updated

TomWoodward committed:
setting type=search for mobile keyboard

TomWoodward committed:
add sidebar animation

Commit:
animateIn search sidebar

Commit:
Merge branch 'search-sidebar' of https://github.com/openstax/rex-web into search-sidebar

Commit:
lint error fixes

TomWoodward committed:
fix horrible error

TomWoodward committed:
fix linter errors

TomWoodward committed:
fix unit tests

TomWoodward committed:
test coverage

TomWoodward committed:
remove unused code

TomWoodward committed:
code coverage

Commit:
fixed margin and padding left
on mobile for search sidebar,
replaced close icon with custom one made for toc,
replacing aria labels with translation msg.

Commit:
snaps updated, lint errors fix

TomWoodward committed:
fix safari search outline

TomWoodward committed:
Merge branch 'master' into search-sidebar

Commit:
input default styles fix for safari

Commit:
Merge branch 'search-sidebar' of https://github.com/openstax/rex-web into search-sidebar

Commit:
adding translation for aria label
in print, updated snap

TomWoodward committed:
move some variables around

Commit:
updated snaps, added opacity to animation for spec to work

TomWoodward committed:
fix unit test

TomWoodward committed:
:camera:

TomWoodward committed:
get sidebar stuff out of app theme

TomWoodward committed:
Merge branch 'search-sidebar' of github.com:openstax/books-web into search-sidebar

KarinaMendez2 commented on src/app/content/components/Toolbar.tsx:
```diff
@@ -116,20 +149,25 @@ const SearchInputWrapper = styled.form`
   color: ${toolbarIconColor.base};
   border: solid 0.1rem;
   border-radius: 0.2rem;
+  width: 38rem;
```
from Devika: 1. SEARCH BOX: a. MVP The search box should be 380px wide.

KarinaMendez2 commented on src/app/content/components/Toolbar.tsx:
```diff
@@ -233,87 +270,161 @@ const TopBar = styled.div`
+`;
+
+// tslint:disable-next-line:variable-name
+const ToggleSeachResultsText = styled.button`
```
you suggested it to be a button before, since it's clickable.

Commit:
renaming input height variable,
changing element types for mouse events

Commit:
Merge branch 'search-sidebar' of https://github.com/openstax/rex-web into search-sidebar

TomWoodward committed:
update sidebar snaps

Commit:
disable search sidebar on print

TomWoodward committed:
Merge branch 'search-sidebar' of github.com:openstax/books-web into search-sidebar

TomWoodward committed:
:camera:

TomWoodward committed:
reset snaps

TomWoodward committed:
:camera:

TomWoodward committed:
:camera:

TomWoodward committed:
working on sidebar animation

TomWoodward committed:
sidebar animation works good

TomWoodward committed:
add toolbar expanded toggle to redux state, rejigger search state

TomWoodward committed:
fix tests

TomWoodward committed:
fix errors in tests

TomWoodward committed:
fix prerenderspecs

TomWoodward committed:
fix prerenderspec

TomWoodward committed:
use links for links

TomWoodward committed:
rename content/sidebar content/TableOfContents

TomWoodward committed:
revert unnecessary change

TomWoodward committed:
don't use toc components from search sidebar

TomWoodward committed:
fix active state for search result details elements

TomWoodward committed:
fix active state for search results

TomWoodward committed:
update search sidebar animation styles

TomWoodward committed:
update search client

TomWoodward committed:
toc icon back in action

TomWoodward committed:
don't close search panel when clicking results on desktop

TomWoodward committed:
unrek scrolling

TomWoodward committed:
:camera:

TomWoodward committed:
fix sidebar close

TomWoodward committed:
:camera:
